include/devicetree.h

```diff
@@ -442,6 +442,39 @@
  */
 #define DT_PROP_LEN(node_id, prop) DT_PROP(node_id, prop##_LEN)
 
+/**
+ * @brief Get a property's logical length or a default value
+ *
+ * Here, "length" is a number of elements, which may differ from the
+ * property's size in bytes.
+ *
+ * The return value depends on the property's type:
+ *
+ * - for types array, string-array, and uint8-array, this expands
+ *   to the number of elements in the array
+ * - for type phandles, this expands to the number of phandles
+ * - for type phandle-array, this expands to the number of
+ *   phandle and specifier blocks in the property
+ * - for a default value if the property does not exists
+ *
+ * These properties are handled as special cases:
+ *
+ * - reg property: use DT_NUM_REGS(node_id) instead
+ * - interrupts property: use DT_NUM_IRQS(node_id) instead
+ *
+ * It is an error to use this macro with the reg or interrupts properties.
+ *
+ * For other properties, behavior is undefined.
```

Comment:
Please follow the example of DT_PROP_OR when writing the docstring by describing the implementation in terms of what it's based on, instead of copy/pasting details that apply to DT_PROP_LEN.

Something like this instead:

```
 * @brief Like DT_PROP_LEN(), but with a fallback to default_value
 *
 * If the node has the property (as determined by DT_NODE_HAS_PROP()),
 * this expands to DT_PROP_LEN(node_id, prop). The default_value
 * parameter is not expanded in this case.
 *
 * Otherwise, this expands to default_value.
```